io.openems.edge.timeofusetariff.entsoe/src/io/openems/edge/timeofusetariff/entsoe/TouEntsoeImpl.java

```diff
@@ -69,82 +66,75 @@ public TouEntsoeImpl() {
-			this.currency = currency;
-			this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
-		};
+		if (config.securityToken() == null) {
+			this.logError(this.log, "Please enter Security token to access ENTSO-E");
+			return;
+		}
+		this.config = config;
 
 		// Trigger to activate when the currency from Meta is updated.
-		this.subscribe(updateCurrency);
+		this.setMeta();
 
-		updateCurrency.accept(this.meta.getCurrencyChannel().getNextValue().asEnum());
+		this.updateCurrency.accept(this.meta.getCurrencyChannel().getNextValue());
 	}
 
 	@Deactivate
 	protected void deactivate() {
 		super.deactivate();
-		this.unsubscribeAll();
+		this.unsetMeta();
 		ThreadPoolUtils.shutdownAndAwaitTermination(this.executor, 0);
 	}
 
 	/**
-	 * Subscribes to the Currency channel to trigger on update.
-	 * 
-	 * @param updateCurrency The callback {@link Consumer}.
+	 * Initializes the Currency channel listener. Adds the callback to the channel.
 	 */
-	private void subscribe(Consumer<Currency> updateCurrency) {
-
-		Consumer<Value<Integer>> subscription = currency -> {
-			updateCurrency.accept(currency.asEnum());
-		};
-		this.subscriptions.add(subscription);
-		this.meta.getCurrencyChannel().onSetNextValue(subscription);
+	private void setMeta() {
+		this.meta.getCurrencyChannel().onSetNextValue(this.updateCurrency);
 	}
 
 	/**
-	 * Unsubscribes from all the Subscriptions.
+	 * Removes the Callback from Currency Channel.
 	 */
-	private void unsubscribeAll() {
-		this.subscriptions.forEach(subscription -> {
-			this.meta.getCurrencyChannel().removeOnSetNextValueCallback(subscription);
-		});
+	private void unsetMeta() {
+		this.meta.getCurrencyChannel().removeOnSetNextValueCallback(this.updateCurrency);
 	}
 
 	private final Runnable task = () -> {
 		var token = this.config.securityToken();
-		var areaCode = this.config.biddingZone().getCode();
+		var areaCode = this.config.biddingZone().code;
 		var fromDate = ZonedDateTime.now().truncatedTo(ChronoUnit.HOURS);
 		var toDate = fromDate.plusDays(1);
 		var unableToUpdatePrices = false;
 
 		try {
-			var result = EntsoeApi.query(token, areaCode, fromDate, toDate);
+			final var result = EntsoeApi.query(token, areaCode, fromDate, toDate);
+			final var entsoeCurrency = Utils.parseCurrency(result);
 
-			final double exchangeRate = this.currency == Currency.EUR //
+			final var exchangeRate = this.currency.toString() == entsoeCurrency //
 					? 1 // No need to fetch exchange rate from API.
 					: Utils.exchangeRateParser(ExchangeRateApi.getExchangeRates(), this.currency);
 
 			// Parse the response for the prices
-			this.prices.set(Utils.parse(result, "PT60M", exchangeRate));
+			this.prices.set(Utils.parsePrices(result, "PT60M", exchangeRate));
 
 			// store the time stamp
 			this.updateTimeStamp = ZonedDateTime.now();
 		} catch (IOException | ParserConfigurationException | SAXException | OpenemsNamedException e) {
+			this.logWarn(this.log, e.getMessage());
```

Comment:
It's still helpful to get a hint like `Unable to Update Entsoe Time-Of-Use Price:`. The message alone adds no additional information, because it is anyway part of `printStackTrace`